core/primitives/src/shard_layout.rs

```diff
@@ -83,14 +98,77 @@ pub struct ShardLayoutV1 {
     /// Maps shards from the last shard layout to shards that it splits to in this shard layout,
     /// Useful for constructing states for the shards.
     /// None for the genesis shard layout
-    shards_split_map: Option<ShardSplitMap>,
+    shards_split_map: Option<ShardsSplitMap>,
     /// Maps shard in this shard layout to their parent shard
     /// Since shard_ids always range from 0 to num_shards - 1, we use vec instead of a hashmap
     to_parent_shard_map: Option<Vec<ShardId>>,
     /// Version of the shard layout, this is useful for uniquely identify the shard layout
     version: ShardVersion,
 }
 
+impl ShardLayoutV1 {
+    // In this shard layout the accounts are divided into ranges, each range is
+    // mapped to a shard. The shards are contiguous and start from 0.
+    fn account_id_to_shard_id(&self, account_id: &AccountId) -> ShardId {
+        let mut shard_id: ShardId = 0;
+        for boundary_account in &self.boundary_accounts {
+            if account_id < boundary_account {
+                break;
+            }
+            shard_id += 1;
+        }
+        shard_id
+    }
+}
+
+/// Making the shard ids non-contiguous.
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ShardLayoutV2 {
+    /// The boundary accounts are the accounts on boundaries between shards.
+    /// Each shard contains a range of accounts from one boundary account to
+    /// another - or the smallest or largest account possible. The total
+    /// number of shards is equal to the number of boundary accounts plus 1.
+    ///
+    /// The shard ids do not need to be contiguous or sorted.
+    boundary_accounts: Vec<AccountId>,
+
+    /// The shard ids corresponding to the shards defined by the boundary
+    /// accounts. The invariant between boundary_accounts and shard_ids is that
+    /// boundary_accounts.len() + 1 == shard_ids.len().
+    ///
+    /// The shard id at index i corresponds to the shard with account range:
+    /// [boundary_accounts[i -1], boundary_accounts[i]).
+    ///
+    shard_ids: Vec<ShardId>,
+
+    /// A mapping from the parent shard to child shards. Maps shards from the
+    /// previous shard layout to shards that they split to in this shard layout.
+    shards_split_map: Option<ShardsSplitMapV2>,
+    /// A mapping from the child shard to the parent shard. Maps shards in this
+    /// shard layout to their parent shards.
+    shards_parent_map: Option<ShardsParentMapV2>,
```

Comment:
Why are the shards_parent_map and shards_split_map optional?

Reply:
It's for the genesis case. It's used in tests mostly.